### DISCUSSION TITLE
Trying to create a `ref` but getting "Object does not exist" "422"

### DISCUSSION THREAD

**Discussion by Vanals** [octokit/octokit.js#2284] [Q&A] [answered]
Hi,

I am at my first experience with Octokit and I am trying to create a pull request to a repo, clicking a button.

First thing first I am creating a `ref` which from what I get is basically another way to say creating a 'branch'?

In the long term, what I am trying to achieve is: 

1) Create a branch to a repo
2) Push some file names, path and content to be added to the repo
3) create the pr.

To satisfy the point 1) I am using the following method

```
    const octokit = new Octokit({
      auth: gitHubAuthToken
    })
    
   // I believe  "cryptoJS.SHA1" argument SHOULD BE "refs/heads/main"? THE BRANCH I WANT TO BRANCH OUT FROM?
    const SHA1 = cryptoJS.SHA1('refs/heads/main')
    const encSHA1 = cryptoJS.enc.Hex.stringify(SHA1)

    console.log(encSHA1)

    await octokit.request('POST /repos/{owner}/{repo}/git/refs', {
      owner: 'nameOfTheOrganizationWhereTheRepoSits,
      repo: 'newskit-themes-publisher',
      ref: 'refs/heads/name-of-branch-i-want-to-create,
      sha: `${SHA1}`
    })
  ```
  
  
  But I am getting 422 -  Object does not exist.
  
  
  Any clue on why I might be getting it?
  
  Thank you for your help and any clarification! Haven't found any solution on the internet, and yes I am following the documentation [here](https://docs.github.com/en/rest/git/refs#create-a-reference)

**Comment by timrogers** [comment, marked as answer]
👋🏻 Hi there!

If you're creating a branch using `POST /repos/{owner}/{repo}/git/refs`, you use the `ref` to set the name of the new branch and the `sha` to point to the Git commit SHA which should be the `HEAD` of that branch.

For example, if I wanted to create a branch called `tim-tim-tim` in `octokit/octokit.js` and point it to commit [`37d6fd6e8f50dc75fb82e2de3e8bb70359a9b578`](https://github.com/octokit/octokit.js/commit/37d6fd6e8f50dc75fb82e2de3e8bb70359a9b578), then I'd use the following code:

```js
await octokit.request('POST /repos/{owner}/{repo}/git/refs', {
  owner: 'octokit',
  repo: 'octokit.js',
  ref: 'refs/heads/tim-tim-tim',
  sha: '37d6fd6e8f50dc75fb82e2de3e8bb70359a9b578'
})
```

Can you give that a go and let me know how you get on?

**Reply by oscard0m** [reply]
I would recommend using [`octokit-plugin-create-pull-request` by @gr2m](https://github.com/gr2m/octokit-plugin-create-pull-request).

It is probably covering most of the problems and edge cases you need.

Here's [an example](https://github.com/oscard0m/octoherd-script-add-cache-to-node-github-action/blob/72e4ad13bd1a8088db62056089d9b0168d004f82/script.js#L70-L98) of how I used it myself for a script.

---

As a side note, if you think your script could be useful for the community, I encourage you to share it at https://github.com/octoherd! (you can ping me there if you need help).

**Reply by Vanals** [reply]
Thank you @timrogers  now makes sense. It wasn't clear in the documentation what the SHA1 had to be of. It worked.

I guess now need a function to retrieve the latest commit SHA1? haven't found one yet.

Lastly I've been trying to delete the ref(branch) using the API but I am being told it doesn't exist, while it does. I can checkout -b to the branch.

I am using the syntax

```
   // BRANCH REF STRING I CREATED PREVIOUSLY
    const branchToCreateRef = 'refs/heads/randomizing-theme-color'
    
    
    await octokit.request('DELETE /repos/{owner}/{repo}/git/refs/{ref}', {
      owner: publisherRepoOwner,
      repo: repoName,
      ref: branchToCreateRef,
    })

```

But I get told it doesn't exist. Is the ref value I am passing correct? In the doc doesn't explain what should be

**Reply by timrogers** [reply]
When you try to delete the branch, it looks like you're repeating "refs" twice because it's in the URL path and your branch variable. 

Could you try changing it so "refs" only appears once? I'm not sure if that's the solution, but it might be. 

I'll check how to get the latest commit and will get back to you.

**Reply by Vanals** [reply]
Thank you TIm, very helpful!

I think your suggestion was right.
Doing

```
    const repoName = "my-repo-name"
    const branchToDelete = 'refs/heads/randomizing-theme-color'
    const publisherRepoOwner = 'newscorp-ghfb'

    const octokit = new Octokit({
      auth: gitHubAuthToken
    })

    //TODO BEFORE USING: REMOVE PR IF IN REPO, DELETE BRANCH IF EXISTING.
    // TODO deleting branch
    await octokit.request('DELETE /repos/{owner}/{repo}/git/{ref}', {
      owner: publisherRepoOwner,
      repo: repoName,
      ref: branchToDelete,
    })
 ```
    
    
The first time worked but then when trying to test it again is giving me 404 :/ .
The ref exist, but now gives me back 404. The first time I tried gave me 204 successfully. Now seems not working anymore but this should be the syntax I used the first time

**Reply by timrogers** [reply]
Great! ✨ I would expect that API to error if you tried to delete a branch that has already been deleted. Did you re-create the branch before running it again?

You can use the `GET /repos/:owner/:repo/commits/:branch` API to find the latest commit SHA, like this:

```
const { data: { sha } } = await octokit.request('GET /repos/{owner}/{repo}/commits/{branch}', {
        owner: publisherRepoOwner,
        repo: repoName,
        // Replace main with the branch you want to check
        branch: 'main'
      });
```

Hope that helps!

**Reply by Vanals** [reply]
Hi @timrogers !

Yes I did recreate the branch, I can checkout on it but I get 404. Left above the code I use, is not working :/ 

Anyway I had a go with the plugin you suggested AND... it's a bless! It should definitely be part of Octokit, makes things 100x easier!
For now I am good will give few other goes and let you know if I have other questions.

Thank you for your help! much appreciated

**Reply by Vanals** [reply]
Hi @timrogers  @oscard0m ! I have one  concern about the createPullRequest plugin I hope you can reply to.

I am creating a pull request and setting `update: true` so the pull request gets updated if existing.
BUT
I noticed the PR gets updated FORCE PUSHING.

![Screenshot 2022-09-16 at 13 25 20](https://user-images.githubusercontent.com/25549355/190638271-59760609-c56a-448d-9e36-8ca80de784c2.png)

Sometimes, force pushing can create problems when pulling down, or with conflicts.
Is there a way to update the PR with a "normal" commit and push? without force pushing?

Thank you!

```
octokit.createPullRequest({
      owner: publisherRepoOwner,
      repo: repoName,
      title: "TEST: Randomizing COLOUR",
      body: "pull request description",
      head: branchToCreateRef,
      base: "main" /* optional: defaults to default branch */,
      update: true /* optional: set to `true` to enable updating existing pull requests */,
      forceFork: false /* optional: force creating fork even when user has write rights */,
      changes: [
        {
          /* optional: if `files` is not passed, an empty commit is created instead */
          files: {
            [themeTokensSetPath]: JSON.stringify(themeTokensSet, null, 2),
          },
          // TODO
          commit:
            "Test commit text, to be branch topic",
        },
      ],
    }).then((pr) => console.log(pr && pr.data.number));
  }
  ```
  
  
  
  LASTLY, in addition, would you have any suggestions for how to test createPullRequest is working good?Ta

**Reply by oscard0m** [reply]
Hi @Vanals,

> Sometimes, force pushing can create problems when pulling down, or with conflicts.
Is there a way to update the PR with a "normal" commit and push? without force pushing?

Definitely, this would be an interesting feature. Checking existing issues opened for that plugin, I think there is an issue already opened for what you are asking for: https://github.com/gr2m/octokit-plugin-create-pull-request/issues/72. I think the best you can do is to post your case there.

---

> LASTLY, in addition, would you have any suggestions for how to test createPullRequest is working good?

The approach I use to apply when testing Octokit is to capture the requests done by Octokit and assure these requests are done with the expected parameters.

In Octokit we use [`nock`](https://github.com/nock/nock) but there are other solutions out there like [`msw`](https://mswjs.io/). 

Feel free to ping me if you would like some assistance on that.

Cheers!

**Comment by Vanals** [comment]
Thank you for the heads up @oscard0m 

I wish this was something implemented and maintained by Octokit?
I am maybe a bit reluctant of using this external package, as it feels a bit an insecure approach?

In general I trie to use packages from trusted sources, used and maintained bigger organizations.

Can you tell me anything that might let me be more sure about using it?
Maybe I am simply not experienced about this plugins

Don't get me wrong, having a look it looks exactly what I need!!

**Reply by timrogers** [reply]
I understand your preference to use prominent packages published by well-known companies. If that's what you're comfortable with, then you should stick with it ❤️

The package that @oscard0m mentioned was created by one of the maintainers of Octokit, @gr2m - but you're right that it isn't part of Octokit itself at the moment. 

In this case, you can review the open source code on GitHub to see what exactly it does, so there shouldn't be any real risk.

**Reply by Vanals** [reply]
I will check with my team anyway! ;)
